crossdock/client/headers/behavior.go

```diff
@@ -0,0 +1,173 @@
+	fatals.NotEmpty(encoding, "encoding is required")
+
+	var caller headerCaller
+
+	switch encoding {
+	case "raw":
+		caller = rawCaller{raw.New(rpc.Channel("yarpc-test"))}
+	case "json":
+		caller = jsonCaller{json.New(rpc.Channel("yarpc-test"))}
+	case "thrift":
+		caller = thriftCaller{echoclient.New(rpc.Channel("yarpc-test"))}
+	default:
+		fatals.True(false, "unknown encoding %q", encoding)
+	}
+
+	token1 := random.String(10)
+	token2 := random.String(10)
+
+	tests := []struct {
+		desc string
+		give transport.Headers
+		want transport.Headers
+	}{
+		{
+			"valid headers",
+			transport.Headers{"token1": token1, "token2": token2},
+			transport.Headers{"token1": token1, "token2": token2},
+		},
+		{
+			"non-string values",
+			transport.Headers{"token": "42"},
+			transport.Headers{"token": "42"},
+		},
+		{
+			"empty strings",
+			transport.Headers{"token": ""},
+			transport.Headers{"token": ""},
+		},
+		{
+			"no headers",
+			nil,
+			transport.Headers{},
+		},
+		{
+			"empty map",
+			transport.Headers{},
+			transport.Headers{},
+		},
+		{
+			"varying casing",
+			transport.Headers{"ToKeN1": token1, "tOkEn2": token2},
+			transport.Headers{"token1": token1, "token2": token2},
+			// TODO(abg): Do we expect to get back the same casing or
+			// normalized?
+		},
+		{
+			"http header conflict",
+			transport.Headers{"Rpc-Procedure": "does not exist"},
+			transport.Headers{"rpc-procedure": "does not exist"},
+			// TODO(abg): Do we expect to get back the same casing or
+			// normalized?
+		},
+	}
+
+	for _, tt := range tests {
+		got, err := caller.Call(tt.give)
+		if checks.NoError(err, "%v: call failed", tt.desc) {
+			assert.Equal(tt.want, got, "%v: returns valid headers", tt.desc)
+		}
+	}
+}
+
+type headerCaller interface {
+	Call(transport.Headers) (transport.Headers, error)
+}
+
+type rawCaller struct{ c raw.Client }
+
+func (c rawCaller) Call(h transport.Headers) (transport.Headers, error) {
+	ctx, _ := context.WithTimeout(context.Background(), time.Second)
+	_, res, err := c.c.Call(&raw.Request{
+		Context:   ctx,
+		Headers:   h,
+		Procedure: "echo/raw",
+		TTL:       time.Second, // TODO context contains the timeout
+	}, []byte("hello"))
+
+	if err != nil {
+		return nil, err
+	}
+	return res.Headers, nil
+}
+
+type jsonCaller struct{ c json.Client }
+
+func (c jsonCaller) Call(h transport.Headers) (transport.Headers, error) {
+	ctx, _ := context.WithTimeout(context.Background(), time.Second)
+
+	var resBody interface{}
+	res, err := c.c.Call(&json.Request{
+		Context:   ctx,
+		Headers:   h,
+		Procedure: "echo",
+		TTL:       time.Second, // TODO context contains the timeout
+	}, map[string]interface{}{}, &resBody)
+
+	if err != nil {
+		return nil, err
+	}
+	return res.Headers, nil
+}
+
+type thriftCaller struct{ c echoclient.Interface }
+
+func (c thriftCaller) Call(h transport.Headers) (transport.Headers, error) {
+	ctx, _ := context.WithTimeout(context.Background(), time.Second)
```

Comment:
nit: if you want to ignore the second value, and have the same timeout for all of these, would it be better to have a private function in crossdock `newTestContext() context.Context`? you could choose to inline that too, and can avoid some of the noise